privval/file_test.go

```diff
@@ -143,118 +167,158 @@ func TestUnmarshalValidatorKey(t *testing.T) {
-	}
-
-	for _, c := range cases {
-		cpb := c.ToProto()
-		err = privVal.SignVote("mychainid", cpb, false)
-		require.Error(t, err, "expected error on signing conflicting vote")
-	}
-
-	// try signing a vote with a different time stamp
-	sig := vote.Signature
-	vote.Timestamp = vote.Timestamp.Add(time.Duration(1000))
-	err = privVal.SignVote("mychainid", v, false)
-	require.NoError(t, err)
-	assert.Equal(sig, vote.Signature)
 }
 
 func TestSignProposal(t *testing.T) {
-	assert := assert.New(t)
-
-	privVal, _, _ := newTestFilePV(t)
-
-	randbytes := cmtrand.Bytes(tmhash.Size)
-	randbytes2 := cmtrand.Bytes(tmhash.Size)
-
-	block1 := types.BlockID{
-		Hash:          randbytes,
-		PartSetHeader: types.PartSetHeader{Total: 5, Hash: randbytes},
-	}
-	block2 := types.BlockID{
-		Hash:          randbytes2,
-		PartSetHeader: types.PartSetHeader{Total: 10, Hash: randbytes2},
+	for _, keyType := range kt.ListSupportedKeyTypes() {
+		t.Run(keyType, func(t *testing.T) {
+			assert := assert.New(t)
+			chainID := "mychainid" + keyType
+
+			keyGenF := func() (crypto.PrivKey, error) {
+				return kt.GenPrivKey(keyType)
+			}
+			privVal, _, _ := newTestFilePV(t, keyGenF)
+
+			randbytes := cmtrand.Bytes(tmhash.Size)
+			randbytes2 := cmtrand.Bytes(tmhash.Size)
+
+			block1 := types.BlockID{
+				Hash:          randbytes,
+				PartSetHeader: types.PartSetHeader{Total: 5, Hash: randbytes},
+			}
+			block2 := types.BlockID{
+				Hash:          randbytes2,
+				PartSetHeader: types.PartSetHeader{Total: 10, Hash: randbytes2},
+			}
+			height, round := int64(10), int32(1)
+
+			// sign a proposal for first time
+			proposal := newProposal(height, round, block1)
+			pbp := proposal.ToProto()
+			err := privVal.SignProposal(chainID, pbp)
+			sig := pbp.Signature
+			require.NoError(t, err, "expected no error signing proposal")
+
+			// try to sign the same proposal again; should be fine
+			err = privVal.SignProposal(chainID, pbp)
+			require.NoError(t, err, "expected no error on signing same proposal")
+
+			// Verify proposal signature
+			pubKey, err := privVal.GetPubKey()
+			require.NoError(t, err)
+			assert.True(pubKey.VerifySignature(types.ProposalSignBytes(chainID, pbp), sig))
+
+			// now try some bad Proposals
+			cases := []*types.Proposal{
+				newProposal(height, round-1, block1),   // round regression
+				newProposal(height-1, round, block1),   // height regression
+				newProposal(height-2, round+4, block1), // height regression and different round
+				newProposal(height, round, block2),     // different block
+			}
+
+			for _, c := range cases {
+				err = privVal.SignProposal(chainID, c.ToProto())
+				require.Error(t, err, "expected error on signing conflicting proposal")
+			}
+
+			// try signing a proposal with a different time stamp
+			proposal.Timestamp = proposal.Timestamp.Add(time.Duration(1000))
+			pbp2 := proposal.ToProto()
+			err = privVal.SignProposal(chainID, pbp2)
+			require.NoError(t, err)
+			assert.Equal(sig, pbp2.Signature)
```

Comment:
Here as well, the changes I made are fixing a bug that was present for a few years (in test code, not production)